File: src/turnStore.ts
import { mkdirSync, readFileSync, writeFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import {
  type ConversationKey,
  type TurnId,
  type TurnRunning,
  type TurnState,
  brandConversationKey,
  brandTurnId,
  isRecord,
  readNumber,
  readString,
} from "./types.js";

export function turnsRoot(): string {
  return process.env.TURNS_DIR ?? "/tmp/turns";
}

export class TurnStore {
  private readonly byKey = new Map<string, TurnState>();
  private readonly byId = new Map<string, TurnState>();
  private spawnCount = 0;

  getSpawnCount(): number {
    return this.spawnCount;
  }

  incrementSpawn(): void {
    this.spawnCount += 1;
  }

  getRunning(key: ConversationKey): TurnRunning | undefined {
    const current = this.byKey.get(key);
    if (current !== undefined && current.kind === "running") {
      return current;
    }
    return undefined;
  }

  getById(turnId: TurnId): TurnState | undefined {
    return this.byId.get(turnId);
  }

  runningCount(): number {
    let n = 0;
    for (const state of this.byKey.values()) {
      if (state.kind === "running") {
        n += 1;
      }
    }
    return n;
  }

  put(state: TurnState): void {
    this.byKey.set(state.conversationKey, state);
    this.byId.set(state.turnId, state);
    persistTurn(state);
  }

  markDone(turnId: TurnId, finalReply?: string): TurnState | undefined {
    const current = this.byId.get(turnId);
    if (current === undefined || current.kind !== "running") {
      return current;
    }
    const done: TurnState = {
      kind: "done",
      turnId: current.turnId,
      conversationKey: current.conversationKey,
      startMessage: current.startMessage,
      followups: current.followups,
      inboundReactions: current.inboundReactions,
      setReactions: current.setReactions,
    };
    if (finalReply !== undefined) {
      done.finalReply = finalReply;
    }
    this.put(done);
    return done;
  }
}

function persistTurn(state: TurnState): void {
  const dir = join(turnsRoot(), state.turnId);
  mkdirSync(dir, { recursive: true });
  writeFileSync(join(dir, "state.json"), JSON.stringify(state, null, 2), "utf8");
}

export function loadPersistedTurn(turnId: TurnId): unknown {
  const path = join(turnsRoot(), turnId, "state.json");
  if (!existsSync(path)) {
    return undefined;
  }
  const raw: unknown = JSON.parse(readFileSync(path, "utf8"));
  return raw;
}

export function newTurnId(key: ConversationKey): TurnId {
  const safe = key.replace(/[^a-zA-Z0-9._-]+/g, "_").slice(0, 48);
  return brandTurnId(`trn_${safe}_${Date.now()}_${Math.random().toString(16).slice(2, 8)}`);
}

export function rehydrateKey(value: string): ConversationKey {
  return brandConversationKey(value);
}

export function rehydrateTurnId(value: string): TurnId {
  return brandTurnId(value);
}

export function readPersistedPid(raw: unknown): number | undefined {
  if (!isRecord(raw)) {
    return undefined;
  }
  return readNumber(raw.pid);
}

export function readPersistedKind(raw: unknown): string | undefined {
  if (!isRecord(raw)) {
    return undefined;
  }
  return readString(raw.kind);
}

File: python/teams_harness/turn_store.py
from __future__ import annotations

import json
import os
import random
import time
from pathlib import Path

from .types import ConversationKey, TurnDone, TurnId, TurnRunning, TurnState, brand_turn_id


def turns_root() -> str:
    return os.environ.get("TURNS_DIR", "/tmp/turns")


class TurnStore:
    def __init__(self) -> None:
        self._by_key: dict[str, TurnState] = {}
        self._by_id: dict[str, TurnState] = {}
        self._spawn_count = 0

    def get_spawn_count(self) -> int:
        return self._spawn_count

    def increment_spawn(self) -> None:
        self._spawn_count += 1

    def get_running(self, key: ConversationKey) -> TurnRunning | None:
        current = self._by_key.get(key)
        if current is not None and current.kind == "running":
            return current
        return None

    def get_by_id(self, turn_id: TurnId) -> TurnState | None:
        return self._by_id.get(turn_id)

    def running_count(self) -> int:
        return sum(1 for state in self._by_key.values() if state.kind == "running")

    def put(self, state: TurnState) -> None:
        self._by_key[state.conversation_key] = state
        self._by_id[state.turn_id] = state
        persist_turn(state)

    def mark_done(self, turn_id: TurnId, final_reply: str | None = None) -> TurnState | None:
        current = self._by_id.get(turn_id)
        if current is None or current.kind != "running":
            return current
        done = TurnDone(
            kind="done",
            turn_id=current.turn_id,
            conversation_key=current.conversation_key,
            start_message=current.start_message,
            followups=list(current.followups),
            inbound_reactions=list(current.inbound_reactions),
            set_reactions=list(current.set_reactions),
            final_reply=final_reply,
        )
        self.put(done)
        return done


def persist_turn(state: TurnState) -> None:
    directory = Path(turns_root()) / state.turn_id
    directory.mkdir(parents=True, exist_ok=True)
    (directory / "state.json").write_text(json.dumps(_state_to_json(state), indent=2), encoding="utf-8")


def new_turn_id(key: ConversationKey) -> TurnId:
    safe = "".join(ch if ch.isalnum() or ch in "._-" else "_" for ch in key)[:48]
    return brand_turn_id(f"trn_{safe}_{int(time.time() * 1000)}_{random.randrange(0, 16**6):x}")


def _state_to_json(state: TurnState) -> dict[str, object]:
    msg = state.start_message
    payload: dict[str, object] = {
        "kind": state.kind,
        "turnId": state.turn_id,
        "conversationKey": state.conversation_key,
        "startMessage": {
            "kind": "message",
            "messageId": msg.message_id,
            "text": msg.text,
            "conversationKey": msg.conversation_key,
            "conversationId": msg.conversation_id,
            "serviceUrl": msg.service_url,
            "fromId": msg.from_id,
            "conversationType": msg.conversation_type,
        },
        "followups": [m.message_id for m in state.followups],
        "inboundReactions": [
            {"messageId": r.message_id, "emoji": r.emoji, "action": r.action, "fromId": r.from_id}
            for r in state.inbound_reactions
        ],
    }
    if state.kind == "running":
        payload["pid"] = state.pid
        payload["startedAt"] = state.started_at
        if state.grok_session_id is not None:
            payload["grokSessionId"] = state.grok_session_id
    if state.kind == "done" and state.final_reply is not None:
        payload["finalReply"] = state.final_reply
    return payload
